File: app/src/main/java/me/blog/korn123/easyphotomap/utils/AsyncUtils.java
package me.blog.korn123.easyphotomap.utils;

import android.graphics.drawable.Drawable;
import android.util.Log;
import android.widget.ImageView;

import org.apache.commons.lang.StringUtils;

import java.net.URI;

import me.blog.korn123.easyphotomap.thumbnail.AsyncDrawable;
import me.blog.korn123.easyphotomap.thumbnail.BitmapWorkerTask;

/**
 * Created by CHO HANJOONG on 2016-08-03.
 */
public class AsyncUtils {

    public static void loadBitmap(ImageView imageView, String imagePath, int widthHeight) {
        if (cancelPotentialWork(imagePath, imageView)) {
            BitmapWorkerTask task = new BitmapWorkerTask(imagePath, imageView);
            final AsyncDrawable asyncDrawable = new AsyncDrawable(task, imagePath);
            imageView.setImageDrawable(asyncDrawable);
            task.execute(imagePath, String.valueOf(widthHeight));
        }
    }

    public static BitmapWorkerTask getBitmapWorkerTask(ImageView imageView) {
        BitmapWorkerTask bitmapWorkerTask = null;
        if (imageView != null) {
            final Drawable drawable = imageView.getDrawable();
            if (drawable instanceof AsyncDrawable) {
                final AsyncDrawable asyncDrawable = (AsyncDrawable) drawable;
                bitmapWorkerTask = asyncDrawable.getBitmapWorkerTask();
            }
        }
        return bitmapWorkerTask;
    }

    public static boolean cancelPotentialWork(String filePath, ImageView imageView) {

        String previousFilepath = "empty";
        BitmapWorkerTask task = null;
        if (imageView.getDrawable() instanceof AsyncDrawable) {
            AsyncDrawable asyncDrawable  = (AsyncDrawable) imageView.getDrawable();
            if (asyncDrawable != null) {
                task = asyncDrawable.getBitmapWorkerTask();
                previousFilepath = task.filePath;
            }
        }

        if (StringUtils.equals(filePath, previousFilepath)) {
//                Log.i("cancelPotentialWork", "bitmapData " + bitmapData);
            // Cancel previous task
            return false;
        } else {
            if (task != null) {
                task.cancel(true);
            }
            // The same work is already in progress
            return true;
        }
    }
}

File: app/src/main/java/me/blog/korn123/easyphotomap/utils/AsyncUtils.kt
package me.blog.korn123.easyphotomap.utils

import android.widget.ImageView
import me.blog.korn123.easyphotomap.thumbnail.AsyncDrawable
import me.blog.korn123.easyphotomap.thumbnail.BitmapWorkerTask
import org.apache.commons.lang.StringUtils

/**
 * Created by CHO HANJOONG on 2016-08-03.
 */
object AsyncUtils {

//    fun loadBitmap(imageView: ImageView, imagePath: String, widthHeight: Int) {
//        if (cancelPotentialWork(imagePath, imageView)) {
//            val task = BitmapWorkerTask(imagePath, imageView, null)
//            val asyncDrawable = AsyncDrawable(task, imagePath)
//            imageView.setImageDrawable(asyncDrawable)
//            task.execute(imagePath, widthHeight.toString())
//        }
//    }

    fun getBitmapWorkerTask(imageView: ImageView?): BitmapWorkerTask? {
        var bitmapWorkerTask: BitmapWorkerTask? = null
        if (imageView != null) {
            val drawable = imageView.drawable
            if (drawable is AsyncDrawable) {
                bitmapWorkerTask = drawable.bitmapWorkerTask
            }
        }
        return bitmapWorkerTask
    }

//    private fun cancelPotentialWork(filePath: String, imageView: ImageView): Boolean {
//        var previousFilepath = "empty"
//        var task: BitmapWorkerTask? = null
//        if (imageView.drawable is AsyncDrawable) {
//            val asyncDrawable = imageView.drawable as AsyncDrawable
//            task = asyncDrawable.bitmapWorkerTask
//            previousFilepath = task.filePath
//        }
//
//        return when (StringUtils.equals(filePath, previousFilepath)) {
//            true -> false
//            false -> {
//                if (task != null) {
//                    task.cancel(true)
//                }
//                true
//            }
//        }
//    }

}
